### PR TITLE
Windows path fix

### PR DESCRIPTION
**Pull request by lospennikov-nbl**

### CODE REVIEW

File: spec/Machine/sample-10-spec.js [hunk 1]
```diff
@@ -19,8 +19,8 @@ describe('Machine', () => {
   });
 
   it('should do something alright #1', () => {
-    expect(machine.execute('@include "input.nut"')).toBe(result);
+    expect(machine.execute('@include "input.nut"').replace(/\\/g, '/')).toBe(result);
```
Comment (sergey-nbl) [inline spec/Machine/sample-10-spec.js]: Is it really need here? result.nut.out doesn't contain any file path

File: spec/Machine/sample-10-spec.js [hunk 2]
```diff
@@ -19,8 +19,8 @@ describe('Machine', () => {
...
     machine.generateLineControlStatements = true;
-    expect(machine.execute('@include "input.nut"')).toBe(resultWithLC);
+    expect(machine.execute('@include "input.nut"').replace(/\\/g, '/')).toBe(resultWithLC);
```
Comment (sergey-nbl) [inline spec/Machine/sample-10-spec.js]: I'd prefer not to have this unobvious code. Is this the only test for windows path? May we generate resultWithLC?

**Review by sergey-nbl — APPROVED** [review]

File: spec/Machine/init.js [hunk 3]
```diff
@@ -26,7 +26,12 @@ module.exports = (sampleFile) => {
     },
 
     getResultWithLineControl: () => {
-      return fs.readFileSync(sampleFile + '.out-lc', 'utf-8');
+      const content = fs.readFileSync(sampleFile + '.out-lc', 'utf-8');
+      // replace for files, that have predefined line control
+      // with char '$' instead of slash for compatibility with Unix and Windows
+      // for example, line control "#line 1 1$2$test" will be
+      // "#line 1 1/2/test" on Unix and "#line 1 1\2\test" on Windows
+      return content.replace(/\$/g, path.sep);
```
Comment (snork-nbl) [inline spec/Machine/init.js]: This is definitely not the right way to do it. If there is a $ symbol anywhere in the file (say in a string or something), it'll be incorrectly replaced.

File: src/Readers/AbstractReader.js [hunk 4]
```diff
@@ -44,8 +44,10 @@ class AbstractReader {
     const __FILE__ = path.basename(source);
     let __PATH__;
 
-    let parsedURL = url.parse(source);
-    if (parsedURL.protocol) {
+    const parsedURL = url.parse(source);
```
Comment (snork-nbl) [inline src/Readers/AbstractReader.js]: The variable is used just once. Why to introduce it?

File: spec/Machine/init.js [hunk 5]
```diff
@@ -26,7 +26,28 @@ module.exports = (sampleFile) => {
     },
 
     getResultWithLineControl: () => {
-      return fs.readFileSync(sampleFile + '.out-lc', 'utf-8');
+      let content;
+      try {
```
Comment (snork-nbl) [inline spec/Machine/init.js:30]: No need to invent a bicycle: https://stackoverflow.com/questions/8683895/how-do-i-determine-the-current-operating-system-with-node-js

File: src/Readers/AbstractReader.js [hunk 6]
```diff
@@ -44,8 +44,8 @@ class AbstractReader {
     const __FILE__ = path.basename(source);
     let __PATH__;
 
-    let parsedURL = url.parse(source);
-    if (parsedURL.protocol) {
+    // url parse can find protocol in Windows-style path, so we check it
+    if (url.parse(source).protocol && !path.parse(source).root) {
```
Comment (snork-nbl) [inline src/Readers/AbstractReader.js]: it's better to cache url.parse(source).protocol than to call it twice

File: spec/Machine/init.js (hunk 5 above)
Comment (lospennikov-nbl) [reply spec/Machine/init.js:30]: Main thing is path sep, not platform.  In my opinion, needless to handle many types of fs, if it has same filepath separator.

Comment (ppetrosh) [inline spec/Machine/init.js]: It's just windows that you need to distinguish from everything else.

Comment (snork-nbl) [reply spec/Machine/init.js:30]: It's just windows that you need to distinguish from everything else.

**Review by sergey-nbl — APPROVED** [review]

**Review by snork-nbl — APPROVED** [review]

**Pull request merged**